pkg/transforms/mqttsecret.go

```diff
@@ -0,0 +1,270 @@
+	opts.SetClientID(mqttConfig.ClientId)
+	opts.SetAutoReconnect(mqttConfig.AutoReconnect)
+	//avoid casing issues
+	mqttConfig.AuthMode = strings.ToLower(mqttConfig.AuthMode)
+	sender := &MQTTSecretSender{
+		client:         nil,
+		mqttConfig:     mqttConfig,
+		persistOnError: persistOnError,
+		opts:           opts,
+	}
+
+	return sender
+}
+func (sender *MQTTSecretSender) getSecrets(edgexcontext *appcontext.Context) (*mqttSecrets, error) {
+	// No Auth? No Problem!...No secrets required.
+	if sender.mqttConfig.AuthMode == AuthModeNone {
+		return nil, nil
+	}
+
+	secrets, err := edgexcontext.GetSecrets(sender.mqttConfig.SecretPath)
+	if err != nil {
+		return nil, err
+	}
+	mqttSecrets := &mqttSecrets{
+		username:     secrets[MQTTSecretUsername],
+		password:     secrets[MQTTSecretPassword],
+		keypemblock:  []byte(secrets[MQTTSecretClientKey]),
+		certpemblock: []byte(secrets[MQTTSecretClientCert]),
+		capemblock:   []byte(secrets[MQTTSecretCACert]),
+	}
+
+	return mqttSecrets, nil
+}
+func (sender *MQTTSecretSender) validateSecrets(secrets mqttSecrets) error {
+	caCertPool := x509.NewCertPool()
+	if sender.mqttConfig.AuthMode == AuthModeUsernamePassword {
+		if secrets.username == "" || secrets.password == "" {
+			return errors.New("AuthModeUsernamePassword selected however username or password was not found at secret path")
+		}
+
+	} else if sender.mqttConfig.AuthMode == AuthModeCert {
+		// need both to make a successful connection
+		if len(secrets.keypemblock) <= 0 || len(secrets.certpemblock) <= 0 {
+			return errors.New("AuthModeCert selected however the key or cert PEM block was not found at secret path")
+		}
+	} else if sender.mqttConfig.AuthMode == AuthModeCA && len(secrets.capemblock) <= 0 {
+		return errors.New("AuthModeCA selected however no PEM Block was found at secret path")
+	}
+
+	if len(secrets.capemblock) > 0 {
+		ok := caCertPool.AppendCertsFromPEM([]byte(secrets.capemblock))
+		if !ok {
+			return errors.New("Error parsing CA Certificate")
+		}
+	}
+	return nil
+}
+func (sender *MQTTSecretSender) configureMQTTClientForAuth(secrets mqttSecrets) error {
+	var cert tls.Certificate
+	var err error
+	caCertPool := x509.NewCertPool()
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: sender.mqttConfig.SkipCertVerify,
+	}
+	switch sender.mqttConfig.AuthMode {
+	case AuthModeUsernamePassword:
+		sender.opts.SetUsername(secrets.username)
+		sender.opts.SetPassword(secrets.password)
+	case AuthModeCert:
+		cert, err = tls.X509KeyPair(secrets.certpemblock, secrets.keypemblock)
+		if err != nil {
+			return err
+		}
+		tlsConfig.Certificates = []tls.Certificate{cert}
+	case AuthModeCA:
+		break
+	case AuthModeNone:
+		return nil
+	default:
+		return nil
+	}
+
+	if len(secrets.capemblock) > 0 {
+		ok := caCertPool.AppendCertsFromPEM(secrets.capemblock)
+		if !ok {
+			return errors.New("Error parsing CA PEM block")
+		}
+		tlsConfig.ClientCAs = caCertPool
+	}
+
+	sender.opts.SetTLSConfig(tlsConfig)
```

Comment:
Do we always want to call SetTLSConfig()? i.e. if I am only using Username/Password we would want to skip call to SetTLSConfig()?

Reply:
It shouldn't hurt, if you didnt provide client certs or ca then its not really setting it to anything since the struct wont have those values populated